### PR TITLE
[FEATURE] Resolve variables in prometheus legend formatter

### PR DESCRIPTION
**Pull request by saminzadeh**
Variables are now supported in the Prometheus legend formatter

image
image

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

## UI Changes

- [x] Changes that impact the UI include screenshots and/or screencasts of the relevant changes.
- [x] Code follows the [UI guidelines](https://github.com/perses/perses/blob/main/ui/ui-guidelines.md).
- [x] Visual tests are stable and unlikely to be flaky. See [Storybook](https://github.com/perses/perses/tree/main/ui/storybook#visual-tests) and [e2e](https://github.com/perses/perses/tree/main/ui/e2e#visual-tests) docs for more details. Common issues include:
  - Is the data inconsistent? You need to mock API requests.
  - Does the time change? You need to use consistent time values or mock time utilities.
  - Does it have loading states? You need to wait for loading to complete.

### CODE REVIEW

File: ui/prometheus-plugin/src/plugins/prometheus-time-series-query/plugin.test.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    end: new Date('01-01-2023'),
+    start: new Date('01-02-2023'),
+  },
+  variableState: {},
+};
+
+describe('PrometheusTimeSeriesQuery', () => {
+  it('should properly resolve variable dependencies', () => {
+    if (!PrometheusTimeSeriesQuery.dependsOn) throw new Error('dependsOn is not defined');
+    const { variables } = PrometheusTimeSeriesQuery.dependsOn(
```
Comment (saminzadeh) [inline ui/prometheus-plugin/src/plugins/prometheus-time-series-query/plugin.test.ts:21]: Trying to write a smoke test but there seems to be an issue with the imports

image
@sjcobb any ideas?

Comment (saminzadeh) [reply ui/prometheus-plugin/src/plugins/prometheus-time-series-query/plugin.test.ts:37]: looks like `jest.mock('echarts/core');` fixed it for now. Need to investigate another way to do this globally for free without the consumer needing to know this

**Comment by juliepagano** [comment]
I don't _think_ this is caused by your PR because I can reproduce it with non-variables too, but the chart in the panel editor seems to weirdly blank out the lines sometimes when you modify the legend name. Is that something you might have time to take a closer look at and/or file a bug for later while you're in here.

I can pretty reliably reproduce this with the top-right "legend example" chart  in the "benchmark" dashboard.

image

**Comment by juliepagano** [comment]
UX feedback for maybe a follow-up, an info tooltip, and/or documentation. I could see this being confusing and/or not very discoverable alongside the handlebars-ish syntax for other interpolation we recommend in the placeholder.

image

**Review by juliepagano — APPROVED** [review]
Left some notes about a bug I maybe saw while testing your PR and some suggestions about UX follow-up, but nothing that's blocking this PR. Code changes LGTM! 🎉

File: ui/prometheus-plugin/src/plugins/prometheus-time-series-query/get-time-series-data.ts [hunk 2]
```diff
@@ -51,6 +51,12 @@ export const getTimeSeriesData: TimeSeriesQueryPlugin<PrometheusTimeSeriesQueryS
   let query = spec.query.replace('$__rate_interval', `15s`);
   query = replaceTemplateVariables(query, context.variableState);
 
+  let seriesNameFormat = spec.series_name_format;
+  // if series name format is defined, replace template variable placeholders in series name format
+  if (seriesNameFormat) {
+    seriesNameFormat = replaceTemplateVariables(seriesNameFormat, context.variableState);
```
Comment (sjcobb) [inline ui/prometheus-plugin/src/plugins/prometheus-time-series-query/get-time-series-data.ts:57]: Nice! LGTM
image

**Review by sjcobb — APPROVED** [review]

**Comment by saminzadeh** [comment]
> I don't _think_ this is caused by your PR because I can reproduce it with non-variables too, but the chart in the panel editor seems to weirdly blank out the lines sometimes when you modify the legend name. Is that something you might have time to take a closer look at and/or file a bug for later while you're in here.
> 
> I can pretty reliably reproduce this with the top-right "legend example" chart in the "benchmark" dashboard.
> 
> image

Ah yeah I noticed that too. Filed a bug: https://github.com/perses/perses/issues/1235

**Pull request merged**